File: apps/mobile/modules/import-sources/android/src/main/java/expo/modules/importsources/CopyRegistry.kt
package expo.modules.importsources

/**
 * Process-wide cancellation registry shared by both module classes, giving
 * `copyToPath` and `copyAsset` one copyId namespace. Completion and cancel
 * race to exactly one outcome: a finished copy ignores a late cancel; a
 * cancelled copy never delivers a result.
 */
class CopyRegistry {
  companion object {
    val shared = CopyRegistry()
  }

  private data class Entry(var cancelled: Boolean = false, var onCancel: (() -> Unit)? = null)

  private val entries = HashMap<String, Entry>()

  @Synchronized
  fun register(copyId: String, onCancel: (() -> Unit)? = null) {
    entries[copyId] = Entry(onCancel = onCancel)
  }

  fun cancel(copyId: String) {
    val action: (() -> Unit)?
    synchronized(this) {
      val entry = entries[copyId] ?: return
      if (entry.cancelled) return
      entry.cancelled = true
      action = entry.onCancel
      entry.onCancel = null
    }
    action?.invoke()
  }

  @Synchronized
  fun isCancelled(copyId: String): Boolean = entries[copyId]?.cancelled ?: false

  /** Removes the entry; returns whether cancel won (suppress the result). */
  @Synchronized
  fun finish(copyId: String): Boolean {
    val wasCancelled = entries[copyId]?.cancelled ?: false
    entries.remove(copyId)
    return wasCancelled
  }
}

File: apps/mobile/modules/import-sources/ios/CopyRegistry.swift
import Foundation

/// Process-wide cancellation registry shared by both module classes, so one
/// copyId namespace covers `copyToPath` and `copyAsset`. Completion and cancel
/// race to exactly one outcome: a copy that already finished ignores a late
/// cancel; a cancelled copy must never deliver a result (the copier checks
/// `isCancelled` between chunks and `finish` reports whether cancel won).
public final class CopyRegistry {
  public static let shared = CopyRegistry()

  private struct Entry {
    var cancelled = false
    var onCancel: (() -> Void)?
  }

  private let lock = NSLock()
  private var entries: [String: Entry] = [:]

  public init() {}

  public func register(_ copyId: String, onCancel: (() -> Void)? = nil) {
    lock.lock()
    defer { lock.unlock() }
    entries[copyId] = Entry(cancelled: false, onCancel: onCancel)
  }

  /// Unknown ids no-op. Fires the entry's cancel action at most once.
  public func cancel(_ copyId: String) {
    lock.lock()
    guard var entry = entries[copyId], !entry.cancelled else {
      lock.unlock()
      return
    }
    entry.cancelled = true
    let action = entry.onCancel
    entry.onCancel = nil
    entries[copyId] = entry
    lock.unlock()
    action?()
  }

  public func isCancelled(_ copyId: String) -> Bool {
    lock.lock()
    defer { lock.unlock() }
    return entries[copyId]?.cancelled ?? false
  }

  /// Removes the entry; returns whether cancel won the race (caller must then
  /// suppress its result).
  @discardableResult
  public func finish(_ copyId: String) -> Bool {
    lock.lock()
    defer { lock.unlock() }
    let wasCancelled = entries[copyId]?.cancelled ?? false
    entries[copyId] = nil
    return wasCancelled
  }
}
